src/boot/efi/boot.c

```diff
@@ -1814,6 +1814,56 @@ static VOID config_entry_add_osx(Config *config) {
         }
 }
 
+
+static VOID config_entry_add_windows(Config *config, EFI_HANDLE *device, EFI_FILE *root_dir) {
+        EFI_STATUS err;
+        UINTN len;
+        const CHAR16 *title = NULL;
+        _cleanup_freepool_ CHAR8 *bcd = NULL;
+
+        if (!config->auto_entries)
+                return;
+
+        /* Try to find a better title. */
+        err = file_read(root_dir, L"\\EFI\\Microsoft\\Boot\\BCD", 0, 100*1024, &bcd, &len);
+        if (!EFI_ERROR(err)) {
+                const CHAR16 prefix[] = L"Windows ";
+                const UINTN prefix_len = sizeof(prefix) - sizeof(CHAR16);
+
+                const struct {
+                        const CHAR16 *title;
+                        const UINTN version_len;
+                } versions[] = {
+                        { L"Windows 11", 6 },
+                        { L"Windows 10", 6 },
+                        { L"Windows 8.1", 8 },
+                        { L"Windows 8", 4, },
+                        { L"Windows 7", 4 },
+                        { L"Windows Vista", 12 },
+                };
+                len -= prefix_len + 12;
+
+                for (CHAR8 *start = bcd; !title && start <= bcd + len; start++) {
+                        if (start[0] != prefix[0] || CompareMem(start, prefix, prefix_len) != 0)
+                                continue;
+
+                        start += prefix_len;
+                        for (UINTN i = 0; i < ELEMENTSOF(versions); i++) {
+                                if (CompareMem(start, versions[i].title + (prefix_len / sizeof(CHAR16)), versions[i].version_len) == 0) {
+                                        title = versions[i].title;
+                                        break;
+                                }
+                        }
+                }
+        }
```

Comment:
So as I understand this is a brute force search of the first 100K of the BCD file? This sounds terribly slow to me.

Isn't there a better way?

don't we have something like memmem() available somewhere in the EFI library apis?

isn't there a better way to detect this? PE metadata of the main windows EFI binary?

Reply:
This is the best I can think of. There is version information in the EFI binary, but it's not in the PE section. So it would amount to inventing a parser for whatever format those are or registry hives (BCD). I don't wanna do either.

In practice, this is not slow at all / you won't notice. I can measure this if you want. Note that the BCD of a fresh install is around 40K only. So we don't even hit the limit usually.

Reply:
ok. (i do acknowledge we look for our own signatures like that too, maybe we should abstract this "search for some strings" thing in a helper function actually)